sky/shell/ui/animator.cc

```diff
@@ -7,151 +7,103 @@
+    RequestFrame();
     return;
   }
 
-  begin_time_ = ftl::TimePoint::Now();
-  ftl::TimePoint frame_time =
-      time_stamp
-          ? ftl::TimePoint() + ftl::TimeDelta::FromMicroseconds(time_stamp)
-          : begin_time_;
-
-  is_ready_to_draw_ = true;
-  engine_->BeginFrame(frame_time);
-  bool was_ready_to_draw = is_ready_to_draw_;
-  is_ready_to_draw_ = false;
-
-  // If we were still ready to draw when done with the frame, that means we
-  // didn't draw anything this frame and we should acknowledge the frame
-  // ourselves instead of waiting for the rasterizer to acknowledge it.
-  if (was_ready_to_draw)
-    OnFrameComplete();
+  blink::Threads::Gpu()->PostTask(
+      [&]() { rasterizer_->Draw(&layer_tree_pipeline_); });
 }
 
 void Animator::Render(std::unique_ptr<flow::LayerTree> layer_tree) {
-  if (!is_ready_to_draw_)
-    return;  // Only draw once per frame.
-  is_ready_to_draw_ = false;
-
-  layer_tree->set_construction_time(ftl::TimePoint::Now() - begin_time_);
-
-  // TODO(abarth): Doesn't this leak if OnFrameComplete never runs?
-  rasterizer_->Draw(
-      reinterpret_cast<uint64_t>(layer_tree.release()),
-      base::Bind(&Animator::OnFrameComplete, weak_factory_.GetWeakPtr()));
+  renderable_tree_ = std::move(layer_tree);
 }
 
-void Animator::OnFrameComplete() {
-  DCHECK(outstanding_requests_ > 0);
-  --outstanding_requests_;
-  TRACE_COUNTER1("flutter", "outstanding_requests_", outstanding_requests_);
-  if (paused_)
+void Animator::RequestFrame() {
+  if (paused_) {
     return;
-
-  if (did_defer_frame_request_) {
-    did_defer_frame_request_ = false;
-
-    if (!AwaitVSync())
-      BeginFrame(0);
   }
-}
 
-bool Animator::AwaitVSync() {
-  if (frame_scheduler_) {
-    frame_scheduler_->ScheduleFrame(
-        base::Bind(&Animator::Animate, weak_factory_.GetWeakPtr()));
-    return true;
-  } else if (vsync_provider_) {
-    vsync_provider_->AwaitVSync(
-        base::Bind(&Animator::BeginFrame, weak_factory_.GetWeakPtr()));
-    return true;
+  if (!pending_frame_sem_.TryWait()) {
+    // Multiple calls to Animator::RequestFrame will still result in a single
+    // request to the VSyncProvider.
+    return;
   }
-  return false;
-}
-
-void Animator::Reset() {
-  weak_factory_.InvalidateWeakPtrs();
 
-  outstanding_requests_ = 0;
-  TRACE_COUNTER1("flutter", "outstanding_requests_", outstanding_requests_);
-  did_defer_frame_request_ = false;
-  engine_requested_frame_ = false;
-  paused_ = false;
+  // The AwaitVSync is going to call us back at the next VSync. However, we want
+  // to be reasonably certain that the UI thread is not in the middle of a
+  // particularly expensive callout. We post the AwaitVSync to run right after
+  // an idle. This does NOT provide a guarantee that the UI thread has not
+  // started an expensive operation right after posting this message however.
+  // To support that, we need edge triggered wakes on VSync.
+
+  blink::Threads::UI()->PostTask([&]() {
```

Comment:
We don't. Also, we need to Post on the semaphore irrespective of the pause state. Fixing.